### ISSUE TITLE
JTReg test fail : java/lang/invoke/VarHandles/accessibility/TestFieldLookupAccessibility.java

### ISSUE THREAD

**Issue by ben-walsh** [eclipse/openj9#3175]
Multiple sub-test failures due to accessibility differences for MethodHandles when using MH_UNREFLECT_SETTER_ACCESSIBLE lookup method ...
```
 109         MH_UNREFLECT_SETTER_ACCESSIBLE() {
 110             Object lookup(MethodHandles.Lookup l, Field f) throws Exception {
 111                 return l.unreflectSetter(cloneAndSetAccessible(f));
 112             }
 113         },
...
 145         static Field cloneAndSetAccessible(Field f) throws Exception {
 146             // Clone to avoid mutating source field
 147             f = f.getDeclaringClass().getDeclaredField(f.getName());
 148             f.setAccessible(true);
 149             return f;
 150         }
```

Log archive : [methodHandlesAccessProb.tar.gz](https://github.com/eclipse/openj9/files/2455861/methodHandlesAccessProb.tar.gz)

Test passes against Hotspot.

**Comment by ben-walsh** [comment]
Test passes using ...
```
openjdk version "11" 2018-09-25
OpenJDK Runtime Environment AdoptOpenJDK (build 11+28-201810141809)
Eclipse OpenJ9 VM AdoptOpenJDK (build master-fa48180f, JRE 11 Linux amd64-64-Bit Compressed References 20181014_61 (JIT enabled, AOT enabled)
OpenJ9   - fa48180f
OMR      - 6f35a3ce
JCL      - 4f9e2ba5c4 based on jdk-11+28)
```
... closing accordingly.

**Comment by pdbain-ibm** [comment]
Failing on the `MH_UNREFLECT_SETTER_ACCESSIBLE` tests.
In one test case doing a lookup from `pkg.A` to the same class, the following fields should be inaccessible:

> f_package_final
> f_package_static_final
> f_private_static_final
> f_private_final
> f_protected_final
> f_protected_static_final
> f_public_static_final
> f_public_final

**Comment by pdbain-ibm** [comment]
For the record via @DanHeidinga :
```If the underlying field is final, the method throws an IllegalAccessException unless setAccessible(true) has succeeded for this Field object and the field is non-static. Setting a final field in this way is meaningful only during deserialization or reconstruction of instances of classes with blank final fields, before they are made available for access by other parts of a program. Use in any other context may have unpredictable effects, including cases in which other parts of a program continue to use the original value of this field.```
http://cr.openjdk.java.net/~iris/se/11/latestSpec/api/java.base/java/lang/reflect/Field.html#set(java.lang.Object,java.lang.Object)

**Comment by pdbain-ibm** [comment]
... though the test expects final static fields to be accessible.

**Comment by andrewcraik** [comment]
@liqunl note the discussion above since it will be relevant to whether we can fold these accesses at compile-time without protecting them with decompilation points.

**Comment by pdbain-ibm** [comment]
Technically, the test is wrong since it doesn't obey the restriction
`setAccessible(true) has succeeded for this Field object and` **the field is non-static**.
Should we appeal the tests which expect `static final` fields to be accessible?

**Comment by andrewcraik** [comment]
It is definitely worth pushing back on any cases where modifications of finals are expected to be reflected at runtime in a test when the spec disallows this behavior IMO. This behavior requires the compiler to add safety checking code and fallback paths that increase code size and can reduce performance.

**Comment by pdbain-ibm** [comment]
@smlambert This jtreg test is testing behaviour contrary to the spec.  Can we open a pull request against the test in question?  I would prefer not to exclude the test.

**Comment by pdbain-ibm** [comment]
Further to 
https://github.com/eclipse/openj9/issues/3175#issuecomment-437377048:  , note that the code doesn't necessarily need to work correctly after setting the field:
`Use in any other context may have unpredictable effects, including cases in which other parts of a program continue to use the original value of this field.`

**Comment by smlambert** [comment]
I believe questions regarding openjdk test material can be raised through the openjdk forums, mailings lists (as @ben-walsh and others in the CL team can do).  Defects can be reported at the [openjdk bug database](http://bugs.openjdk.java.net )...  typically, before a new issue raised, we should be checking if this is already reported as a bug, some guidance here: https://github.com/AdoptOpenJDK/openjdk-tests/wiki/Guidance-for-Creating-OpenJDK-Test-Defects

**Comment by adamfarley** [comment]
Adam Farley and Andrew Leonard are Authors, not Comitters. 

Authors can:
- Create Bug reports for use by the OpenJDK community
- Modify existing bug reports
- Upload files to those bug reports
- Advise on how to contribute code for review by the OpenJDK community.

Authors cannot:
- Integrate new code into OpenJDK without a committer's approval. 

See here for a guide on contributing code to OpenJDK:
http://openjdk.java.net/contribute/

**Comment by pdbain-ibm** [comment]
@adamfarley or @andrew-m-leonard would you kindly check if a bug against this issue has been opened?

**Comment by adamfarley** [comment]
OpenJDK bugs can be searched for on Google. Waiting for one of us to free up time will block you unnecessarily.

**Comment by adamfarley** [comment]
Googled it for you anyway. The only OpenJDK bug that references TestFieldLookupAccessibility is:
https://bugs.openjdk.java.net/browse/JDK-8153963

**Comment by pdbain-ibm** [comment]
That doesn't cover this issue (static final fields): it is for access tests across module boundaries.

**Comment by smlambert** [comment]
Yes, I was NOT suggesting Adam or Andrew look up the defect, since anyone can (as I thought was indicated by the wiki link I shared), apologies for not being more clear.

I was suggesting if there needed to be either a defect raised or a PR submitted, that Adam or Andrew could potentially help.

**Comment by pdbain-ibm** [comment]
OpenJDK does not in fact allow static final fields to be changed.  Changing a field via `field.set()` throws an `IllegalAccessException` even if the field has been set accessible; `Lookup.unreflect()` on a static final field throws no error, but calls to the resulting setter silently fail.

```
package experimental11;

import java.io.PrintStream;
import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.MethodHandles.Lookup;
import java.lang.reflect.Field;

public class TestStaticFinal {

	static final String staticFinal = "initial";
	static final PrintStream err = System.err;
	static final String fieldName = "staticFinal";
	public static void main(String[] args) {
		try {
			Lookup l = MethodHandles.lookup();
			Field f = TestStaticFinal.class.getDeclaredField(fieldName);
			trySetField(f, "value1");
			err.println("set "+fieldName+" accessible");
			f.setAccessible(true);
			trySetField(f, "value2");
			err.println("unreflect "+fieldName);
			String value = "value3";
			trySetter(f, l, value);
		} catch (Throwable e) {}
	}

	private static void trySetField(Field f, String value) {
		try {
			err.println("set "+fieldName+" to \""+value+"\"");
			f.set(TestStaticFinal.class, value);
		} catch (IllegalAccessException e) {
			err.println("! Exception thrown: "+e.getMessage());
		}
		try {
			err.println(fieldName+" = \""+f.get(f.getDeclaringClass())+"\"");
		} catch (IllegalArgumentException | IllegalAccessException e) {}
	}

	private static void trySetter(Field f, Lookup l, String value) throws Throwable {
		try {
			MethodHandle s = l.unreflectSetter(f);
			err.println("call setter on "+fieldName+" with \""+value+"\"");
			s.invoke(value);
		} catch (IllegalAccessException e) {
			err.println("! Exception thrown: "+e.getMessage());
		}
		try {
			err.println(fieldName+" (via setter)= \""+f.get(f.getDeclaringClass())+"\"");
		} catch (IllegalArgumentException | IllegalAccessException e) {}
		err.println("staticFinal (directly) = \""+staticFinal+"\"");
	}
}
```

Running this we get:

> set staticFinal to "value1"
> ! Exception thrown: Can not set static final java.lang.String field experimental11.TestStaticFinal.staticFinal to java.lang.String
> staticFinal = "initial"
> set staticFinal accessible
> set staticFinal to "value2"
> ! Exception thrown: Can not set static final java.lang.String field experimental11.TestStaticFinal.staticFinal to java.lang.String
> staticFinal = "initial"
> unreflect staticFinal
> call setter on staticFinal with "value3"
> staticFinal (via setter)= "value3"
> staticFinal (directly) = "initial"
> 


Note two things:

1. The setter throws no exception, even though the access is illegal.
2. The setter changes the `Field` object, but not the actual static field.

**Comment by pdbain-ibm** [comment]
I believe this is a bug in OpenJDK's access checking. OpenJDK issue: https://github.ibm.com/runtimes/openjdk-contribution/issues/682

**Comment by smlambert** [comment]
https://github.com/AdoptOpenJDK/openjdk-tests/pull/622
